addons/l10n_in_edi/models/einvoice_api.py

```diff
@@ -0,0 +1,100 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from datetime import timedelta
+
+from odoo import fields, models, api, _
+from odoo.exceptions import AccessError
+
+from odoo.addons.iap import jsonrpc
+
+import logging
+
+_logger = logging.getLogger(__name__)
+
+DEFAULT_IAP_ENDPOINT = "https://l10n-in-edi.api.odoo.com"
+DEFAULT_IAP_TEST_ENDPOINT = "https://l10n-in-edi-proxy-demo.odoo.com"
+
+
+class L10nInEdiWebService(models.AbstractModel):
+    _name = "l10n.in.edi.web.service"
+    _description = "E-invoice API (IN)"
+
+    @api.model
+    def _get_token(self, company):
+        if company._l10n_in_edi_token_is_valid():
+            return company.l10n_in_edi_token
+        elif company.l10n_in_edi_username and company.sudo().l10n_in_edi_password:
+            self._authenticate(company)
+            return company.l10n_in_edi_token
+        return False
+
+    @api.model
+    def _connect_to_server(self, company, url_path, params):
+        user_token = self.env["iap.account"].get("l10n_in_edi")
+        params.update(
+            {
+                "account_token": user_token.account_token,
+                "dbuuid": self.env["ir.config_parameter"].sudo().get_param("database.uuid"),
+                "username": company.l10n_in_edi_username,
+                "gstin": company.vat,
+            }
+        )
+        if company.l10n_in_edi_test_env:
+            default_endpoint = DEFAULT_IAP_TEST_ENDPOINT
+        else:
+            default_endpoint = DEFAULT_IAP_ENDPOINT
+        endpoint = self.env["ir.config_parameter"].sudo().get_param("l10n_in_edi.endpoint", default_endpoint)
+        url = "%s%s" % (endpoint, url_path)
+        try:
+            return jsonrpc(url, params=params, timeout=25)
+        except AccessError as e:
+            _logger.warning("Connection error: %s", e.args[0])
+            return {
+                "error": [
+                    {
+                        "code": "404",
+                        "message": _("Unable to connect to the online E-invoice service. The web service may be temporary down. Please try again in a moment."),
+                    }
+                ]
+            }
+
+    @api.model
+    def _authenticate(self, company):
+        params = {"password": company.sudo().l10n_in_edi_password}
+        response = self._connect_to_server(company, url_path="/iap/l10n_in_edi/1/authenticate", params=params)
+        # validity data-time in Indian standard time(UTC+05:30) so remove that gap and store in odoo
+        if "data" in response:
+            company.l10n_in_edi_token_validity = fields.Datetime.to_datetime(
+                response["data"]["TokenExpiry"]
+            ) - timedelta(hours=5, minutes=30, seconds=00)
```

Comment:
I wonder if there is no nicer way with timezone conversion...

Reply:
```
fields.context_timestamp(
                self.with_context(tz="Asia/Kolkata"),
                fields.Datetime.to_datetime(response["data"]["TokenExpiry"]
            ))
```
this is another way to convert it using the time zone name
what do you suggest?

Reply:
If it works, the timezone is more explicit and better

Reply:
What we have
```
company.l10n_in_edi_token_validity = fields.Datetime.to_datetime(
                response["data"]["TokenExpiry"]
            ) - timedelta(hours=5, minutes=30, seconds=00)
```
Other way to do this using timezone
```
local_time = fields.Datetime.context_timestamp(
                self.with_context(tz="Asia/Kolkata"),
                fields.Datetime.to_datetime(response["data"]["TokenExpiry"]
            ))
            utc_time = local_time.astimezone(pytz.utc)
            company.l10n_in_edi_token_validity = fields.Datetime.to_string(utc_time)
```